tiledb/api/src/array/mod.rs

```diff
@@ -1019,4 +1130,102 @@ pub mod tests {
 
         test_uri.close().map_err(|e| Error::Other(e.to_string()))
     }
+
+    fn create_simple_dense(
+        test_uri: &dyn TestArrayUri,
+        ctx: &Context,
+    ) -> TileDBResult<String> {
+        let domain = {
+            let rows = DimensionBuilder::new(
+                ctx,
+                "id",
+                Datatype::Int32,
+                ([1, 410], 10),
+            )?
+            .build();
+
+            DomainBuilder::new(ctx)?.add_dimension(rows)?.build()
+        };
+
+        let schema = SchemaBuilder::new(ctx, ArrayType::Dense, domain)?
+            .add_attribute(
+                AttributeBuilder::new(ctx, "a", Datatype::Int32)?.build(),
+            )?
+            .build()?;
+
+        let array_uri = test_uri
+            .with_path("quickstart_dense")
+            .map_err(|e| Error::Other(e.to_string()))?;
+        Array::create(ctx, &array_uri, schema)?;
+        Ok(array_uri)
+    }
+
+    fn write_dense_vector_4_fragments(
+        ctx: &Context,
+        array_uri: &str,
+        timestamp: u64,
+    ) -> TileDBResult<()> {
+        // Subarray boundaries.
+        let boundaries = [0, 200, 250, 310, 410];
+
+        for i in 0..4 {
+            // Prepare cell buffer.
+            let low_bound = boundaries[i];
+            let high_bound = boundaries[i + 1];
+
+            let mut data: Vec<i32> = vec![];
+            for j in low_bound..high_bound {
+                data.push(j);
+            }
+
+            let opener = ArrayOpener::new(ctx, array_uri, Mode::Write)?
+                .end_timestamp(timestamp + i as u64 + 1)?;
```

Comment:
I think the timestamps are epoch times, not versions, so `timestamp + i` is probably not going to be what you want.

For writing an array I'm not sure what the open timestamp actually does though.  What is your intent?

Reply:
I am following what ever this test is doing: https://github.com/TileDB-Inc/TileDB/blob/2a7935e46575e6d1dfaed800ade44fdef15cfc0d/test/src/unit-capi-consolidation.cc#L5091

To your specific question those are the lines in the test which adds the timestamps: 
https://github.com/TileDB-Inc/TileDB/blob/2a7935e46575e6d1dfaed800ade44fdef15cfc0d/test/src/unit-capi-consolidation.cc#L717
https://github.com/TileDB-Inc/TileDB/blob/2a7935e46575e6d1dfaed800ade44fdef15cfc0d/test/src/unit-capi-consolidation.cc#L765
https://github.com/TileDB-Inc/TileDB/blob/2a7935e46575e6d1dfaed800ade44fdef15cfc0d/test/src/unit-capi-consolidation.cc#L808
https://github.com/TileDB-Inc/TileDB/blob/2a7935e46575e6d1dfaed800ade44fdef15cfc0d/test/src/unit-capi-consolidation.cc#L852

I have simply refactored in a loop.

Reply:
Hm, I think that test is most likely wrong.  Maybe at some point in the past it was a version but then changed to a timestamp?

I did check one of our other tests which uses the timestamp ranges, and the data on the fragments is definitely epoch time. Or some other choice of many-digit number :)

Or possibly it's benign.  As I alluded I have no idea what the timestamp range does for a *write* - I'm guessing if you just remove this it will do the same thing but be less confusing.

Reply:
I can speak more to it on the call this afternoon, but the basic idea is this: Every write happens at a certain time. This time is encoded into the fragment URIs in the form of `__t1_t2_uuid_v` [(Code here)](https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/storage_format/uri/generate_uri.cc#L41-L65) These tests that are operating at the fragment level will often specify their timestamps manually so that various assertions can be made about how fragment handling works.

However, if you see tests that are doing this but not making specific fragment assertions it can also be a historical artifact due to timing bugs. The timestamps in our filenames are millisecond resolution, which meant that any two fragments written in the same millisecond would be ordered randomly based on the UUID component. This has since been fixed so that any two writes in the same OS process will now be correctly ordered regardless of whether they happen in the same millisecond.

A quick skim of these tests suggests this is probably a reflection of the historical issues around timing since we're not attempting to do fancier consolidation stuff. Though I'm pretty sure there are tests that do that somewhere that would still need them (i.e., if this test were consolidating the two middle fragments we'd need to manually specify timestamps so that they'd be selectable).

Reply:
I'm not following what exactly the proposal is here? What should I be correcting wrt above implementation?

Reply:
Per standup discussion will leave as above.